src/data/batch_generator.h

```diff
@@ -242,13 +253,13 @@ class BatchGenerator : public RNGEngine {
       ABORT_IF(!futureBufferedBatches_.valid(), "Attempted to wait for futureBufferedBatches_ when none pending.\n"
           "This error often occurs when Marian tries to restore the training data iterator, but the corpus has been changed or replaced.\n"
           "If you have changed the training corpus, add --no-restore-corpus to the training command and run it again.");
+
       bufferedBatches_ = std::move(futureBufferedBatches_.get());
-      // if bg thread returns an empty swath, we hit the end of the epoch
-      if (bufferedBatches_.empty()) {
+      if (bufferedBatches_.empty() // i.e., end of Epoch
+          || getSignalFlag(SIGTERM)) { // process received SIGTERM, abandon ship ...
```

Comment:
I see the flag tested at three places. Are these mutually exclusive code paths, or does one check lead to returning dummy batches, and then later on, it is checked again to know to ignore the dummy values?

Reply:
It's a nested loop that you want to break out of as soon as possible, so you need to check the conditions for the inner loops and the outer loop. Keep in mind that SIGTERM may occur at any time. The default signal handler for SIGTERM kills all threads in the process immediately. We don't want that, so that we can save the current state of training before exiting. That means that we have to check for SIGTERM in strategic locations where it is convenient for us to check, preferably in the middle of expensive loops. You also have no guarantee that bufferedBatches is empty after SIGTERM -- it may occur between the assignment and and the check.